drivers/serial/uart_nrfx_uart.c

```diff
@@ -180,13 +207,25 @@ static int uart_nrfx_fifo_fill(struct device *dev,
 {
 	u8_t num_tx = 0;
 
+	uart_sw_event_txrdy_pending = 0;
+
 	while ((len - num_tx > 0) &&
-	       nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)) {
+	       (nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY) ||
+	        uart_sw_event_txrdy)) {
```

Comment:
This (together with these two changes above and one below) becomes hard to maintain and consequently, error prone. I suggest to introduce two functions as follows:
```
static bool event_txdrdy_check(void)
{
	return (nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
		|| uart_sw_event_txdrdy
#endif
	       );
}

static void event_txdrdy_clear(void)
{
	nrf_uart_event_clear(NRF_UART0, NRF_UART_EVENT_TXDRDY);
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
	uart_sw_event_txdrdy = 0;
#endif
}
```
and just use them everywhere in this file instead of `nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)` and `nrf_uart_event_clear(NRF_UART0, NRF_UART_EVENT_TXDRDY)`, respectively.

And in the comment before the `uart_sw_event_txdrdy` variable definition you can just write that this variable is used to simulate the TXDRDY event before the first byte is sent, so that this event can be used as an indication of the transmitter status.